ledger/src/blockstore_processor.rs

```diff
@@ -415,21 +423,33 @@ fn process_batches(
 
 fn schedule_batches_for_execution(
     bank: &BankWithScheduler,
-    batches: &[TransactionBatchWithIndexes<SanitizedTransaction>],
+    locked_entries: impl Iterator<Item = LockedTransactionsWithIndexes<SanitizedTransaction>>,
 ) -> Result<()> {
-    for TransactionBatchWithIndexes {
-        batch,
-        transaction_indexes,
-    } in batches
+    // Track the first error encountered in the loop below, if any.
+    // This error will be propagated to the replay stage, or Ok(()).
+    let mut first_err = Ok(());
+
+    for LockedTransactionsWithIndexes {
+        lock_results,
+        transactions,
+        starting_index,
+    } in locked_entries
     {
-        bank.schedule_transaction_executions(
-            batch
-                .sanitized_transactions()
-                .iter()
-                .zip(transaction_indexes.iter()),
-        )?;
+        // unlock before sending to scheduler.
+        bank.unlock_accounts(transactions.iter().zip(lock_results.iter()));
+        // give ownership to scheduler. capture the first error, but continue the loop
+        // to unlock.
+        // scheduling is skipped if we have already detected an error in this loop
+        first_err = first_err.and_then(|()| {
+            bank.schedule_transaction_executions(
+                transactions
+                    .into_iter()
+                    .enumerate()
+                    .map(|(index, tx)| (tx, index + starting_index)),
```

Comment:
nit: i prefer simpler construct like this:

```patch
@@ -441,11 +441,11 @@ fn schedule_batches_for_execution(
         // to unlock.
         // scheduling is skipped if we have already detected an error in this loop
         first_err = first_err.and_then(|()| {
+            let indexes = starting_index..starting_index+transactions.len();
             bank.schedule_transaction_executions(
                 transactions
                     .into_iter()
-                    .enumerate()
-                    .map(|(index, tx)| (tx, index + starting_index)),
+                    .zip_eq(indexes),
             )
         });
     }
```

Reply:
061bdbdc17db4f997e7c74774f4794384f46af74